lib/client/keystore.go

```diff
@@ -578,8 +582,29 @@ func (fs *fsLocalNonSessionKeyStore) kubeCertPath(idx KeyIndex, kubename string)
 	return keypaths.KubeCertPath(fs.KeyDir, idx.ProxyHost, idx.Username, idx.ClusterName, kubename)
 }
 
+// acquireFileLock is trying to lock the file, until it's successful or timeout is exceeded.
+// File will be created if it doesn't exist.
+func acquireFileLock(filePath string, timeout time.Duration) (func() error, error) {
+	fileLock := flock.New(filePath)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	if _, err := fileLock.TryLockContext(ctx, 10*time.Millisecond); err != nil {
+		return nil, err
+	}
+
+	return fileLock.Unlock, nil
+}
+
 // AddKnownHostKeys adds a new entry to `known_hosts` file.
 func (fs *fsLocalNonSessionKeyStore) AddKnownHostKeys(hostname, proxyHost string, hostKeys []ssh.PublicKey) (retErr error) {
+	// We're trying to serialize our writes to the 'known_hosts' file to avoid corruption, since there
+	// are cases when multiple tsh instances will try to write to it.
+	unlock, err := acquireFileLock(fs.knownHostsPath(), 5*time.Second)
+	if err != nil {
+		return trace.WrapWithMessage(err, "could not acquire lock for the `known_hosts` file")
+	}
+	defer utils.StoreErrorOf(unlock, &retErr)
```

Comment:
I don't see `retErr` is ever used or returned.

Reply:
I get the part it's aggregating the errors. But `retErr` is never consumed by anyone (I don't see any return using it)

Reply:
Yep, what @tigrato said :) Just a way to collect errors from defered stuff.

Reply:
I really don't mind leaving it like this. But it is not true that `no error is lost in the process`. We collect them but we are not checking them or logging them.

Reply:
@greedy52 not sure what you mean. It's checked by `func (tc *TeleportClient) UpdateKnownHosts()` ( [in lib/client/api.go](https://github.com/gravitational/teleport/blob/master/lib/client/api.go#L3664) ) and by tests.

Reply:
oh my bad. i missed the defer capture part. Sorry guys. Ignore this.